### DISCUSSION TITLE
#586 discussion on using ClientCog over commands.Cog

### DISCUSSION THREAD

**Discussion by ooliver1** [nextcord/nextcord#588] [Ideas]
discussion for #586 as to not clutter the main PR page

**Comment by MaskDuck** [comment]
imo this leads to several problem.

1. this make the code uglier.
new one:
```py
from somefile import SomeCog
from someotherfile import SomeOtherCog
from someotherotherfile import SomeOtherOtherCog

...


bot.add_cog(SomeCog(bot))
bot.add_cog(SomeOtherCog(bot))
bot.add_cog(SomeOtherOtherCog(bot))
```
old one:
```py
exts = ['somecog', 'someothercog', 'someotherothercog']
for ext in exts:
  bot.load_extension(ext)
```

**Reply by ToasterUwU** [reply]
There are definitely better ways to do that. Like putting all files with Cogs into a folder with a `__init__.py` which imports it for you and then just use one line of import. After that, you can also just loop over it as you did with the old example.

**Comment by alentoghostflame** [comment]
@DenverCoder1 mentioned me, and while I kinda wanted to talk about this on the Discord server, I might as well say it here.

While I'm not super against telling people to use `ClientCog` instead `commands.Cog`, I'm not that much for it. `ClientCog` was my workaround for adding basic application command cog support to both `Client` and `commands.Bot` because IMO, cogs are an essential feature to the library itself, and shouldn't be just an extension. However, `ClientCog` at this time is JUST a workaround, JUST for application commands. It does not support listeners/events, or `cog_unload()`, or have the few extra bits that `commands.Cog` has.

I will agree with the above comment by @MaskDuck (this makes code uglier), but that's mostly because `Client` doesn't support extensions, which IMO it should.

I also have to mention that people will be looking at this, see `ClientCog`, will use it instead of `commands.Cog`, and be confused about why listeners/whatever don't work with it. Once more features are moved from `commands.Cog` to `ClientCog`, I'll be much happier about this change. Until then, I'm kinda meh about it.

Quick Edit: However, I do like using the main lib in examples, rather than `ext.*` stuff, as that's partially regarded as "not part of the library" and thus shouldn't be in examples (except for examples that specifically show how to use that ext, such as this one kinda?)

**Reply by MaskDuck** [reply]
isnt cog an library extension already

**Reply by MaskDuck** [reply]
like the extension provides shit that isn't already provided by discord

**Reply by MaskDuck** [reply]
cog isnt something already natively provided by discord

**Reply by MaskDuck** [reply]
that's why i agree with this comment. if we are using `ClientCog` then we are misleading cogs' purpose.

**Reply by ToasterUwU** [reply]
I wasn't aware that listeners are not included in the ClientCog... That changes that a lot.

I agree now, that changing the example to ClientCog doesn't make sense right now, but for the library to have a consistent structure, it shouldn't use stuff from an extension while not using the rest of the extension in any way. I think it should be cleaned up, to make the structure and logic consistent, but I'm sure that's already the case for V3.

Until this structure problem is fixed, the "old" example should stay.

@alentoghostflame @ooliver1 @MaskDuck

**Reply by ooliver1** [reply]
#393 I believe is the issue for this
